### PR TITLE
Add cartesian to polar conversion functions

### LINKED ISSUE #7263: Proposal: Adding cartesian to polar conversion functions
*Bugsquad edit:* Repurposing this issue based on https://github.com/godotengine/godot/pull/7285#issuecomment-272994312

@tagcup wrote:
> Conversion between degrees and a general 2D vector doesn't make any sense to me, in mathematical terms: the norm of a 2D vector isn't necessarily equal to 1. You can't convert between two degrees of freedom (Vector2) and one degree of freedom (an angle) in a meaningful way.
> 
> I would rather propose adding functions that converts between polar and cartesian coordinates in 2D: (x,y) <-> (r,θ) where θ is in radians. The API could be something like cartesian_to_polar(x,y) and polar_to_cartesian(r,θ), each returning a vector.
> 
> Or alternatively, simply a polar constructor for Vector2.

--------------

*Original proposal:*

I find that a lot of users (me included) have the need to convert angles (in radian or degrees) into a Vector2 direction.

The current way to get a Vector2 direction from an angle;
```
  var radian_value = PI # 180 degrees
  var my_vec2 = **Vector2(cos(radian_value), sin(radian_value))**
```

I propose that we add new functions such as **rad2vec**(rad_value) and **deg2vec**(deg_value) with the other functions found at the [core](http://docs.godotengine.org/en/stable/classes/class_@gdscript.html).

This will make the process much cleaner for everyone, and we wont have to know/remember this formula. :)

### PR DESCRIPTION
**Pull request by tato**
*Bugsquad edit*: Fixes #7263

An attempt to implement the revision of #7263 . Hopefully I did everything right for my first time. Edit: I didn't. I forgot to configure my editor to use tabs. Any way to change the commit now?

### CODE REVIEW

File: modules/gdscript/gd_functions.cpp [hunk 1]
```diff
@@ -405,6 +407,22 @@ void GDFunctions::call(Function p_func, const Variant **p_args, int p_arg_count,
 			VALIDATE_ARG_NUM(0);
 			r_ret = Math::db2linear((double)*p_args[0]);
 		} break;
+		case MATH_POLAR_TO_CARTESIAN: {
+			VALIDATE_ARG_COUNT(2);
+			VALIDATE_ARG_NUM(0);
+            VALIDATE_ARG_NUM(1);
```
Comment (Noshyaar) [inline modules/gdscript/gd_functions.cpp]: please use tabs

File: modules/visual_script/visual_script_builtin_funcs.cpp [hunk 2]
```diff
@@ -76,6 +76,8 @@ const char *VisualScriptBuiltinFunc::func_name[VisualScriptBuiltinFunc::FUNC_MAX
 	"rad2deg",
 	"linear2db",
 	"db2linear",
+    "polar_to_cartesian",
```
Comment (Noshyaar) [inline modules/visual_script/visual_script_builtin_funcs.cpp]: please use tabs

**Comment by Noshyaar** [comment]
I think in \@GDScript built-in, the convention is to use `2` for `to`, like `polar2cartesian`?

**Comment by ptato** [comment]
> I think in @gdscript built-in, the convention is to use `2` for `to`, like `polar2cartesian`?
True, that name would be more consistent with the rest of the functions

**Comment by Chaosus** [comment]
Hmm, my [PR](https://github.com/godotengine/godot/pull/12035) do changes to gd_functions.cpp/h as well, and im not so familiar with git. There is no conflicts ? Maybe merge my PR first @Noshyaar ?

**Comment by n-pigeon** [comment]
@Chaosus 
If two PRs modify the same file and one is merged to master first, the second will have merge conflicts. GitHub will inform it in PR statistics on the bottom and author of second PR will be asked to fix conflicts :)
It's normal with large projects, happened to my PRs also ;)

**Comment by akien-mga** [comment]
>  If two PRs modify the same file and one is merged to master first, the second will have merge conflicts. GitHub will inform it in PR statistics on the bottom and author of second PR will be asked to fix conflicts :)

Not necessarily, there are merge conflicts only if the PRs modify the same lines. If they modify different parts of the file, git can handle it just fine.

**Comment by ptato** [comment]
I added an enum I had forgotten in `visual_script_builtin_funcs.cpp`, and changed the name of the functions to `polar2cartesian` and `cartesian2polar`. Personally I think saying `x2y` is more confusing than `x_to_y`, but consistency is important. I'm going to try to add documentation and figure out how to add them to mono.

**Comment by ptato** [comment]
1. I added the functions to mono on `Mathf.cs`. I don't think I need to register like in the C++ classes, right?
2. In `VisualScriptBuiltinFunc.xml`, every enum has their integer value hardcoded. In order to keep all the math functions together, I added 2 to the values below the new functions. I hope that is acceptable.

File: doc/classes/VisualScriptBuiltinFunc.xml [hunk 3]
```diff
@@ -147,64 +147,70 @@
 		<constant name="MATH_DB2LINEAR" value="37">
 			Convert the input from decibel volume to linear volume.
 		</constant>
-		<constant name="LOGIC_MAX" value="38">
+		<constant name="MATH_POLAR2CARTESIAN" value="38">
+			Convert the input from decibel volume to linear volume.
```
Comment (Noshyaar) [inline doc/classes/VisualScriptBuiltinFunc.xml]: 🤔

**Comment by akien-mga** [comment]
Why close it?

**Comment by ptato** [comment]
whoops I don't know how I did that

**Comment by ptato** [comment]
How do I reopen it? Or do I create a new one? 😅

**Comment by akien-mga** [comment]
Well weirdly it can't be reopened it seems, I guess you modified your master branch and this PR lost the reference to your commit.

I'd suggest that you put your commit in a different branch and make a new PR, so that if you rebase your master branch in the future, it won't affect the PR.

**Pull request closed without merging**